tests/net/ipv4_fragment/src/main.c

```diff
@@ -0,0 +1,886 @@
+	zassert_mem_equal(hdr->dst, &orig_hdr[offsetof(struct net_ipv4_hdr, dst)],
+			  NET_IPV4_ADDR_SIZE, "IPv4 header destination IP mismatch");
+
+	pkt_len = ntohs(hdr->len);
+	pkt_offset = ntohs(*((uint16_t *)&hdr->offset));
+	pkt_flags = (pkt_offset & ~NET_IPV4_FRAGH_OFFSET_MASK) >> 13;
+	pkt_offset &= NET_IPV4_FRAGH_OFFSET_MASK;
+	pkt_offset *= 8;
+
+	if (final) {
+		zassert_equal(pkt_flags, 0, "IPv4 header fragment flags mismatch");
+	} else {
+		zassert_equal(pkt_flags, BIT(0), "IPv4 header fragment flags mismatch");
+	}
+
+	zassert_equal(net_pkt_get_len(pkt), pkt_len, "IPv4 header length mismatch");
+	zassert_equal(pkt_offset, current_length, "IPv4 header length mismatch");
+	zassert_equal(net_calc_chksum_ipv4(pkt), 0, "IPv4 header checksum mismatch");
+}
+
+static int sender_iface(const struct device *dev, struct net_pkt *pkt)
+{
+	struct net_pkt *recv_pkt;
+	int ret;
+	uint8_t buf_ip_src[4];
+	uint8_t buf_ip_dst[4];
+	uint16_t buf_port_src;
+	uint16_t buf_port_dst;
+	uint16_t offset;
+
+	if (!pkt->buffer) {
+		LOG_ERR("No data to send!");
+		return -ENODATA;
+	}
+
+	if (net_pkt_get_len(pkt) > NET_IPV4_MTU) {
+		LOG_DBG("Too large for test");
+		pkt_recv_size = net_pkt_get_len(pkt);
+		return -EMSGSIZE;
+	}
+
+	if (test_started) {
+		++lower_layer_packet_count;
+		lower_layer_total_size += net_pkt_get_len(pkt);
+
+		/* Verify the fragments */
+		bool last_packet;
+
+		net_pkt_cursor_init(pkt);
+
+		last_packet = ((pkt_recv_size + net_pkt_get_len(pkt)) >= pkt_recv_expected_size ?
+			      true : false);
+		check_ipv4_fragment_header(pkt, (active_test == TEST_UDP ? ipv4_udp :
+					   (active_test == TEST_TCP ? ipv4_tcp :
+					   ipv4_icmp_reassembly_time)), 0, pkt_recv_size,
+					   last_packet);
+		pkt_recv_size += net_pkt_get_len(pkt) - NET_IPV4H_LEN;
+
+		if (last_packet) {
+			last_packet_received = true;
+		}
+
+		if (active_test == TEST_SINGLE_FRAGMENT) {
+			/* Verify that this is an ICMPv4 response */
+			zassert_mem_equal(ipv4_icmp_reassembly_time, pkt->buffer->data,
+					  sizeof(ipv4_icmp_reassembly_time), "Expected ICMP error");
+			goto no_duplicate;
+		}
+
+		/* Duplicate the packet and flip the source/destination IPs and ports then
+		 * re-insert the packets back into the same interface
+		 */
+		recv_pkt = net_pkt_rx_clone(pkt, K_NO_WAIT);
+		net_pkt_set_overwrite(recv_pkt, true);
+
+		/* Read IPs */
+		net_pkt_cursor_init(recv_pkt);
+		net_pkt_skip(recv_pkt, 12);
+		net_pkt_read(recv_pkt, buf_ip_src, sizeof(buf_ip_src));
+		net_pkt_read(recv_pkt, buf_ip_dst, sizeof(buf_ip_dst));
+
+		/* Write opposite IPs */
+		net_pkt_cursor_init(recv_pkt);
+		net_pkt_skip(recv_pkt, 12);
+		net_pkt_write(recv_pkt, buf_ip_dst, sizeof(buf_ip_dst));
+		net_pkt_write(recv_pkt, buf_ip_src, sizeof(buf_ip_src));
+
+		/* Get offset to check if this is the first packet */
+		net_pkt_cursor_init(recv_pkt);
+		net_pkt_skip(recv_pkt, 6);
+		net_pkt_read_be16(recv_pkt, &offset);
+		offset &= 0b1111111111111;
```

Comment:
Hmm, this does not look like a standard C but is only available as a gcc extension. I suggest that normal hex value is used here instead.

Reply:
Fixed